// <auto-generated/>
// Contents of: hl7.fhir.r3.core version: 3.0.2
// Minimum TypeScript Version: 3.7
// FHIR Resource: Sequence

import * as fhir from '../fhir.js';

// @ts-ignore
import { ChromosomeHumanCodes,  ChromosomeHumanCodeType } from '../fhirValueSets/ChromosomeHumanCodes.js';
// @ts-ignore
import { ChromosomeHumanVsValidation } from '../fhirValueSets/ChromosomeHumanVsValidation.js';
// @ts-ignore
import { QualityTypeCodes,  QualityTypeCodeType } from '../fhirValueSets/QualityTypeCodes.js';
// @ts-ignore
import { QualityTypeVsValidation } from '../fhirValueSets/QualityTypeVsValidation.js';
// @ts-ignore
import { RepositoryTypeCodes,  RepositoryTypeCodeType } from '../fhirValueSets/RepositoryTypeCodes.js';
// @ts-ignore
import { RepositoryTypeVsValidation } from '../fhirValueSets/RepositoryTypeVsValidation.js';
// @ts-ignore
import { SequenceTypeCodes,  SequenceTypeCodeType } from '../fhirValueSets/SequenceTypeCodes.js';
// @ts-ignore
import { SequenceTypeVsValidation } from '../fhirValueSets/SequenceTypeVsValidation.js';
/**
 * Valid arguments for the SequenceReferenceSeq type.
 */
export interface SequenceReferenceSeqArgs extends fhir.BackboneElementArgs {
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  chromosome?: fhir.CodeableConceptArgs|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  genomeBuild?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.referenceSeq.genomeBuild
   */
  _genomeBuild?:fhir.FhirElementArgs;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the Sequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  referenceSeqId?: fhir.CodeableConceptArgs|undefined;
  /**
   * A Pointer to another Sequence entity as reference sequence.
   */
  referenceSeqPointer?: fhir.ReferenceArgs|undefined;
  /**
   * A string like "ACGT".
   */
  referenceSeqString?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.referenceSeq.referenceSeqString
   */
  _referenceSeqString?:fhir.FhirElementArgs;
  /**
   * Directionality of DNA sequence. Available values are "1" for the plus strand (5' to 3')/Watson/Sense/positive  and "-1" for the minus strand(3' to 5')/Crick/Antisense/negative.
   */
  strand?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.referenceSeq.strand
   */
  _strand?:fhir.FhirElementArgs;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  windowStart: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.referenceSeq.windowStart
   */
  _windowStart?:fhir.FhirElementArgs;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  windowEnd: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.referenceSeq.windowEnd
   */
  _windowEnd?:fhir.FhirElementArgs;
}

/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export class SequenceReferenceSeq extends fhir.BackboneElement {
  /**
   * Mapping of this datatype to a FHIR equivalent
   */
  public static override readonly _fts_dataType:string = 'SequenceReferenceSeq';
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  public chromosome?: fhir.CodeableConcept|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  public genomeBuild?: fhir.FhirString|undefined;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the Sequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  public referenceSeqId?: fhir.CodeableConcept|undefined;
  /**
   * A Pointer to another Sequence entity as reference sequence.
   */
  public referenceSeqPointer?: fhir.Reference|undefined;
  /**
   * A string like "ACGT".
   */
  public referenceSeqString?: fhir.FhirString|undefined;
  /**
   * Directionality of DNA sequence. Available values are "1" for the plus strand (5' to 3')/Watson/Sense/positive  and "-1" for the minus strand(3' to 5')/Crick/Antisense/negative.
   */
  public strand?: fhir.FhirInteger|undefined;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public windowStart: fhir.FhirInteger|null;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public windowEnd: fhir.FhirInteger|null;
  /**
   * Default constructor for SequenceReferenceSeq - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<SequenceReferenceSeqArgs> = {}, options:fhir.FhirConstructorOptions = {}) {
    super(source, options);
    if (source['chromosome']) { this.chromosome = new fhir.CodeableConcept(source.chromosome); }
    if (source['genomeBuild']) { this.genomeBuild = new fhir.FhirString({value: source.genomeBuild}); }
    if (source['_genomeBuild']) {
      if (this.genomeBuild) { this.genomeBuild.addExtendedProperties(source._genomeBuild!); }
      else { this.genomeBuild = new fhir.FhirString(source._genomeBuild as Partial<fhir.FhirStringArgs>); }
    }
    if (source['referenceSeqId']) { this.referenceSeqId = new fhir.CodeableConcept(source.referenceSeqId); }
    if (source['referenceSeqPointer']) { this.referenceSeqPointer = new fhir.Reference(source.referenceSeqPointer); }
    if (source['referenceSeqString']) { this.referenceSeqString = new fhir.FhirString({value: source.referenceSeqString}); }
    if (source['_referenceSeqString']) {
      if (this.referenceSeqString) { this.referenceSeqString.addExtendedProperties(source._referenceSeqString!); }
      else { this.referenceSeqString = new fhir.FhirString(source._referenceSeqString as Partial<fhir.FhirStringArgs>); }
    }
    if (source['strand']) { this.strand = new fhir.FhirInteger({value: source.strand}); }
    if (source['_strand']) {
      if (this.strand) { this.strand.addExtendedProperties(source._strand!); }
      else { this.strand = new fhir.FhirInteger(source._strand as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['windowStart']) { this.windowStart = new fhir.FhirInteger({value: source.windowStart}); }
    else { this.windowStart = null; }
    if (source['_windowStart']) {
      if (this.windowStart) { this.windowStart.addExtendedProperties(source._windowStart!); }
      else { this.windowStart = new fhir.FhirInteger(source._windowStart as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['windowEnd']) { this.windowEnd = new fhir.FhirInteger({value: source.windowEnd}); }
    else { this.windowEnd = null; }
    if (source['_windowEnd']) {
      if (this.windowEnd) { this.windowEnd.addExtendedProperties(source._windowEnd!); }
      else { this.windowEnd = new fhir.FhirInteger(source._windowEnd as Partial<fhir.FhirIntegerArgs>); }
    }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation(exp:string = ''):fhir.FtsIssue[] {
    let issues:fhir.FtsIssue[] = super.doModelValidation(exp);
    if (exp === '') { exp = 'Sequence.referenceSeq' }
    this.vOS('chromosome',exp)
    this.vOS('genomeBuild',exp)
    this.vOS('referenceSeqId',exp)
    this.vOS('referenceSeqPointer',exp)
    this.vOS('referenceSeqString',exp)
    this.vOS('strand',exp)
    this.vRS('windowStart',exp)
    this.vRS('windowEnd',exp)
    return issues;
  }
}
/**
 * Valid arguments for the SequenceVariant type.
 */
export interface SequenceVariantArgs extends fhir.BackboneElementArgs {
  /**
   * Start position of the variant on the  reference sequence.If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.variant.start
   */
  _start?:fhir.FhirElementArgs;
  /**
   * End position of the variant on the reference sequence.If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.variant.end
   */
  _end?:fhir.FhirElementArgs;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  observedAllele?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.variant.observedAllele
   */
  _observedAllele?:fhir.FhirElementArgs;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  referenceAllele?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.variant.referenceAllele
   */
  _referenceAllele?:fhir.FhirElementArgs;
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  cigar?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.variant.cigar
   */
  _cigar?:fhir.FhirElementArgs;
  /**
   * A pointer to an Observation containing variant information.
   */
  variantPointer?: fhir.ReferenceArgs|undefined;
}

/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export class SequenceVariant extends fhir.BackboneElement {
  /**
   * Mapping of this datatype to a FHIR equivalent
   */
  public static override readonly _fts_dataType:string = 'SequenceVariant';
  /**
   * Start position of the variant on the  reference sequence.If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: fhir.FhirInteger|undefined;
  /**
   * End position of the variant on the reference sequence.If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: fhir.FhirInteger|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public observedAllele?: fhir.FhirString|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public referenceAllele?: fhir.FhirString|undefined;
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  public cigar?: fhir.FhirString|undefined;
  /**
   * A pointer to an Observation containing variant information.
   */
  public variantPointer?: fhir.Reference|undefined;
  /**
   * Default constructor for SequenceVariant - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<SequenceVariantArgs> = {}, options:fhir.FhirConstructorOptions = {}) {
    super(source, options);
    if (source['start']) { this.start = new fhir.FhirInteger({value: source.start}); }
    if (source['_start']) {
      if (this.start) { this.start.addExtendedProperties(source._start!); }
      else { this.start = new fhir.FhirInteger(source._start as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['end']) { this.end = new fhir.FhirInteger({value: source.end}); }
    if (source['_end']) {
      if (this.end) { this.end.addExtendedProperties(source._end!); }
      else { this.end = new fhir.FhirInteger(source._end as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['observedAllele']) { this.observedAllele = new fhir.FhirString({value: source.observedAllele}); }
    if (source['_observedAllele']) {
      if (this.observedAllele) { this.observedAllele.addExtendedProperties(source._observedAllele!); }
      else { this.observedAllele = new fhir.FhirString(source._observedAllele as Partial<fhir.FhirStringArgs>); }
    }
    if (source['referenceAllele']) { this.referenceAllele = new fhir.FhirString({value: source.referenceAllele}); }
    if (source['_referenceAllele']) {
      if (this.referenceAllele) { this.referenceAllele.addExtendedProperties(source._referenceAllele!); }
      else { this.referenceAllele = new fhir.FhirString(source._referenceAllele as Partial<fhir.FhirStringArgs>); }
    }
    if (source['cigar']) { this.cigar = new fhir.FhirString({value: source.cigar}); }
    if (source['_cigar']) {
      if (this.cigar) { this.cigar.addExtendedProperties(source._cigar!); }
      else { this.cigar = new fhir.FhirString(source._cigar as Partial<fhir.FhirStringArgs>); }
    }
    if (source['variantPointer']) { this.variantPointer = new fhir.Reference(source.variantPointer); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation(exp:string = ''):fhir.FtsIssue[] {
    let issues:fhir.FtsIssue[] = super.doModelValidation(exp);
    if (exp === '') { exp = 'Sequence.variant' }
    this.vOS('start',exp)
    this.vOS('end',exp)
    this.vOS('observedAllele',exp)
    this.vOS('referenceAllele',exp)
    this.vOS('cigar',exp)
    this.vOS('variantPointer',exp)
    return issues;
  }
}
/**
 * Valid arguments for the SequenceQuality type.
 */
export interface SequenceQualityArgs extends fhir.BackboneElementArgs {
  /**
   * INDEL / SNP / Undefined variant.
   */
  type: fhir.FhirCode<QualityTypeCodeType>|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.type
   */
  _type?:fhir.FhirElementArgs;
  /**
   * Gold standard sequence used for comparing against.
   */
  standardSequence?: fhir.CodeableConceptArgs|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.start
   */
  _start?:fhir.FhirElementArgs;
  /**
   * End position of the sequence.If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.end
   */
  _end?:fhir.FhirElementArgs;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  score?: fhir.QuantityArgs|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  method?: fhir.CodeableConceptArgs|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  truthTP?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.truthTP
   */
  _truthTP?:fhir.FhirElementArgs;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  queryTP?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.queryTP
   */
  _queryTP?:fhir.FhirElementArgs;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  truthFN?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.truthFN
   */
  _truthFN?:fhir.FhirElementArgs;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  queryFP?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.queryFP
   */
  _queryFP?:fhir.FhirElementArgs;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  gtFP?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.gtFP
   */
  _gtFP?:fhir.FhirElementArgs;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  precision?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.precision
   */
  _precision?:fhir.FhirElementArgs;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  recall?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.recall
   */
  _recall?:fhir.FhirElementArgs;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  fScore?: fhir.FhirDecimal|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.quality.fScore
   */
  _fScore?:fhir.FhirElementArgs;
}

/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export class SequenceQuality extends fhir.BackboneElement {
  /**
   * Mapping of this datatype to a FHIR equivalent
   */
  public static override readonly _fts_dataType:string = 'SequenceQuality';
  /**
   * INDEL / SNP / Undefined variant.
   */
  public type: fhir.FhirCode<QualityTypeCodeType>|null;
  /**
   * Gold standard sequence used for comparing against.
   */
  public standardSequence?: fhir.CodeableConcept|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: fhir.FhirInteger|undefined;
  /**
   * End position of the sequence.If the coordinate system is 0-based then end is is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: fhir.FhirInteger|undefined;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  public score?: fhir.Quantity|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  public method?: fhir.CodeableConcept|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public truthTP?: fhir.FhirDecimal|undefined;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public queryTP?: fhir.FhirDecimal|undefined;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  public truthFN?: fhir.FhirDecimal|undefined;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  public queryFP?: fhir.FhirDecimal|undefined;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  public gtFP?: fhir.FhirDecimal|undefined;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  public precision?: fhir.FhirDecimal|undefined;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  public recall?: fhir.FhirDecimal|undefined;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  public fScore?: fhir.FhirDecimal|undefined;
  /**
   * Default constructor for SequenceQuality - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<SequenceQualityArgs> = {}, options:fhir.FhirConstructorOptions = {}) {
    super(source, options);
    if (source['type']) { this.type = new fhir.FhirCode<QualityTypeCodeType>({value: source.type}); }
    else { this.type = null; }
    if (source['_type']) {
      if (this.type) { this.type.addExtendedProperties(source._type!); }
      else { this.type = new fhir.FhirCode<QualityTypeCodeType>(source._type as Partial<fhir.FhirCode>); }
    }
    if (source['standardSequence']) { this.standardSequence = new fhir.CodeableConcept(source.standardSequence); }
    if (source['start']) { this.start = new fhir.FhirInteger({value: source.start}); }
    if (source['_start']) {
      if (this.start) { this.start.addExtendedProperties(source._start!); }
      else { this.start = new fhir.FhirInteger(source._start as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['end']) { this.end = new fhir.FhirInteger({value: source.end}); }
    if (source['_end']) {
      if (this.end) { this.end.addExtendedProperties(source._end!); }
      else { this.end = new fhir.FhirInteger(source._end as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['score']) { this.score = new fhir.Quantity(source.score); }
    if (source['method']) { this.method = new fhir.CodeableConcept(source.method); }
    if (source['truthTP']) { this.truthTP = new fhir.FhirDecimal({value: source.truthTP}); }
    if (source['_truthTP']) {
      if (this.truthTP) { this.truthTP.addExtendedProperties(source._truthTP!); }
      else { this.truthTP = new fhir.FhirDecimal(source._truthTP as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['queryTP']) { this.queryTP = new fhir.FhirDecimal({value: source.queryTP}); }
    if (source['_queryTP']) {
      if (this.queryTP) { this.queryTP.addExtendedProperties(source._queryTP!); }
      else { this.queryTP = new fhir.FhirDecimal(source._queryTP as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['truthFN']) { this.truthFN = new fhir.FhirDecimal({value: source.truthFN}); }
    if (source['_truthFN']) {
      if (this.truthFN) { this.truthFN.addExtendedProperties(source._truthFN!); }
      else { this.truthFN = new fhir.FhirDecimal(source._truthFN as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['queryFP']) { this.queryFP = new fhir.FhirDecimal({value: source.queryFP}); }
    if (source['_queryFP']) {
      if (this.queryFP) { this.queryFP.addExtendedProperties(source._queryFP!); }
      else { this.queryFP = new fhir.FhirDecimal(source._queryFP as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['gtFP']) { this.gtFP = new fhir.FhirDecimal({value: source.gtFP}); }
    if (source['_gtFP']) {
      if (this.gtFP) { this.gtFP.addExtendedProperties(source._gtFP!); }
      else { this.gtFP = new fhir.FhirDecimal(source._gtFP as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['precision']) { this.precision = new fhir.FhirDecimal({value: source.precision}); }
    if (source['_precision']) {
      if (this.precision) { this.precision.addExtendedProperties(source._precision!); }
      else { this.precision = new fhir.FhirDecimal(source._precision as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['recall']) { this.recall = new fhir.FhirDecimal({value: source.recall}); }
    if (source['_recall']) {
      if (this.recall) { this.recall.addExtendedProperties(source._recall!); }
      else { this.recall = new fhir.FhirDecimal(source._recall as Partial<fhir.FhirDecimalArgs>); }
    }
    if (source['fScore']) { this.fScore = new fhir.FhirDecimal({value: source.fScore}); }
    if (source['_fScore']) {
      if (this.fScore) { this.fScore.addExtendedProperties(source._fScore!); }
      else { this.fScore = new fhir.FhirDecimal(source._fScore as Partial<fhir.FhirDecimalArgs>); }
    }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation(exp:string = ''):fhir.FtsIssue[] {
    let issues:fhir.FtsIssue[] = super.doModelValidation(exp);
    if (exp === '') { exp = 'Sequence.quality' }
    this.vRSV('type',exp,'QualityType',QualityTypeVsValidation,'r')
    this.vOS('standardSequence',exp)
    this.vOS('start',exp)
    this.vOS('end',exp)
    this.vOS('score',exp)
    this.vOS('method',exp)
    this.vOS('truthTP',exp)
    this.vOS('queryTP',exp)
    this.vOS('truthFN',exp)
    this.vOS('queryFP',exp)
    this.vOS('gtFP',exp)
    this.vOS('precision',exp)
    this.vOS('recall',exp)
    this.vOS('fScore',exp)
    return issues;
  }
}
/**
 * Valid arguments for the SequenceRepository type.
 */
export interface SequenceRepositoryArgs extends fhir.BackboneElementArgs {
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  type: fhir.FhirCode<RepositoryTypeCodeType>|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.type
   */
  _type?:fhir.FhirElementArgs;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  url?: fhir.FhirUri|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.url
   */
  _url?:fhir.FhirElementArgs;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  name?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.name
   */
  _name?:fhir.FhirElementArgs;
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  datasetId?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.datasetId
   */
  _datasetId?:fhir.FhirElementArgs;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  variantsetId?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.variantsetId
   */
  _variantsetId?:fhir.FhirElementArgs;
  /**
   * Id of the read in this external repository.
   */
  readsetId?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.repository.readsetId
   */
  _readsetId?:fhir.FhirElementArgs;
}

/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export class SequenceRepository extends fhir.BackboneElement {
  /**
   * Mapping of this datatype to a FHIR equivalent
   */
  public static override readonly _fts_dataType:string = 'SequenceRepository';
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  public type: fhir.FhirCode<RepositoryTypeCodeType>|null;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public url?: fhir.FhirUri|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public name?: fhir.FhirString|undefined;
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  public datasetId?: fhir.FhirString|undefined;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  public variantsetId?: fhir.FhirString|undefined;
  /**
   * Id of the read in this external repository.
   */
  public readsetId?: fhir.FhirString|undefined;
  /**
   * Default constructor for SequenceRepository - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<SequenceRepositoryArgs> = {}, options:fhir.FhirConstructorOptions = {}) {
    super(source, options);
    if (source['type']) { this.type = new fhir.FhirCode<RepositoryTypeCodeType>({value: source.type}); }
    else { this.type = null; }
    if (source['_type']) {
      if (this.type) { this.type.addExtendedProperties(source._type!); }
      else { this.type = new fhir.FhirCode<RepositoryTypeCodeType>(source._type as Partial<fhir.FhirCode>); }
    }
    if (source['url']) { this.url = new fhir.FhirUri({value: source.url}); }
    if (source['_url']) {
      if (this.url) { this.url.addExtendedProperties(source._url!); }
      else { this.url = new fhir.FhirUri(source._url as Partial<fhir.FhirUriArgs>); }
    }
    if (source['name']) { this.name = new fhir.FhirString({value: source.name}); }
    if (source['_name']) {
      if (this.name) { this.name.addExtendedProperties(source._name!); }
      else { this.name = new fhir.FhirString(source._name as Partial<fhir.FhirStringArgs>); }
    }
    if (source['datasetId']) { this.datasetId = new fhir.FhirString({value: source.datasetId}); }
    if (source['_datasetId']) {
      if (this.datasetId) { this.datasetId.addExtendedProperties(source._datasetId!); }
      else { this.datasetId = new fhir.FhirString(source._datasetId as Partial<fhir.FhirStringArgs>); }
    }
    if (source['variantsetId']) { this.variantsetId = new fhir.FhirString({value: source.variantsetId}); }
    if (source['_variantsetId']) {
      if (this.variantsetId) { this.variantsetId.addExtendedProperties(source._variantsetId!); }
      else { this.variantsetId = new fhir.FhirString(source._variantsetId as Partial<fhir.FhirStringArgs>); }
    }
    if (source['readsetId']) { this.readsetId = new fhir.FhirString({value: source.readsetId}); }
    if (source['_readsetId']) {
      if (this.readsetId) { this.readsetId.addExtendedProperties(source._readsetId!); }
      else { this.readsetId = new fhir.FhirString(source._readsetId as Partial<fhir.FhirStringArgs>); }
    }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation(exp:string = ''):fhir.FtsIssue[] {
    let issues:fhir.FtsIssue[] = super.doModelValidation(exp);
    if (exp === '') { exp = 'Sequence.repository' }
    this.vRSV('type',exp,'RepositoryType',RepositoryTypeVsValidation,'r')
    this.vOS('url',exp)
    this.vOS('name',exp)
    this.vOS('datasetId',exp)
    this.vOS('variantsetId',exp)
    this.vOS('readsetId',exp)
    return issues;
  }
}
/**
 * Valid arguments for the Sequence type.
 */
export interface SequenceArgs extends fhir.DomainResourceArgs {
  /**
   * Resource Type Name
   */
  resourceType: "Sequence"|undefined;
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  identifier?: fhir.IdentifierArgs[]|undefined;
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  type?: fhir.FhirCode|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.type
   */
  _type?:fhir.FhirElementArgs;
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  coordinateSystem: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.coordinateSystem
   */
  _coordinateSystem?:fhir.FhirElementArgs;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  patient?: fhir.ReferenceArgs|undefined;
  /**
   * Specimen used for sequencing.
   */
  specimen?: fhir.ReferenceArgs|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  device?: fhir.ReferenceArgs|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  performer?: fhir.ReferenceArgs|undefined;
  /**
   * The number of copies of the seqeunce of interest. (RNASeq).
   */
  quantity?: fhir.QuantityArgs|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  referenceSeq?: fhir.SequenceReferenceSeqArgs|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  variant?: fhir.SequenceVariantArgs[]|undefined;
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall starts from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  observedSeq?: fhir.FhirString|string|undefined;
  /**
   * Extended properties for primitive element: Sequence.observedSeq
   */
  _observedSeq?:fhir.FhirElementArgs;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  quality?: fhir.SequenceQualityArgs[]|undefined;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  readCoverage?: fhir.FhirInteger|number|undefined;
  /**
   * Extended properties for primitive element: Sequence.readCoverage
   */
  _readCoverage?:fhir.FhirElementArgs;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  repository?: fhir.SequenceRepositoryArgs[]|undefined;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  pointer?: fhir.ReferenceArgs[]|undefined;
}

/**
 * Raw data describing a biological sequence.
 */
export class Sequence extends fhir.DomainResource {
  /**
   * Mapping of this datatype to a FHIR equivalent
   */
  public static override readonly _fts_dataType:string = 'Sequence';
  /**
   * Resource Type Name
   */
  public override resourceType: "Sequence";
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  public identifier: fhir.Identifier[];
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  public type?: fhir.FhirCode|undefined;
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  public coordinateSystem: fhir.FhirInteger|null;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  public patient?: fhir.Reference|undefined;
  /**
   * Specimen used for sequencing.
   */
  public specimen?: fhir.Reference|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  public device?: fhir.Reference|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  public performer?: fhir.Reference|undefined;
  /**
   * The number of copies of the seqeunce of interest. (RNASeq).
   */
  public quantity?: fhir.Quantity|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  public referenceSeq?: fhir.SequenceReferenceSeq|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  public variant: fhir.SequenceVariant[];
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall starts from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  public observedSeq?: fhir.FhirString|undefined;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  public quality: fhir.SequenceQuality[];
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  public readCoverage?: fhir.FhirInteger|undefined;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  public repository: fhir.SequenceRepository[];
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  public pointer: fhir.Reference[];
  /**
   * Default constructor for Sequence - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<SequenceArgs> = {}, options:fhir.FhirConstructorOptions = {}) {
    super(source, options);
    this.resourceType = 'Sequence';
    if (source['identifier']) { this.identifier = source.identifier.map((x) => new fhir.Identifier(x)); }
    else { this.identifier = []; }
    if (source['type']) { this.type = new fhir.FhirCode({value: source.type}); }
    if (source['_type']) {
      if (this.type) { this.type.addExtendedProperties(source._type!); }
      else { this.type = new fhir.FhirCode(source._type as Partial<fhir.FhirCodeArgs>); }
    }
    if (source['coordinateSystem']) { this.coordinateSystem = new fhir.FhirInteger({value: source.coordinateSystem}); }
    else { this.coordinateSystem = null; }
    if (source['_coordinateSystem']) {
      if (this.coordinateSystem) { this.coordinateSystem.addExtendedProperties(source._coordinateSystem!); }
      else { this.coordinateSystem = new fhir.FhirInteger(source._coordinateSystem as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['patient']) { this.patient = new fhir.Reference(source.patient); }
    if (source['specimen']) { this.specimen = new fhir.Reference(source.specimen); }
    if (source['device']) { this.device = new fhir.Reference(source.device); }
    if (source['performer']) { this.performer = new fhir.Reference(source.performer); }
    if (source['quantity']) { this.quantity = new fhir.Quantity(source.quantity); }
    if (source['referenceSeq']) { this.referenceSeq = new fhir.SequenceReferenceSeq(source.referenceSeq); }
    if (source['variant']) { this.variant = source.variant.map((x) => new fhir.SequenceVariant(x)); }
    else { this.variant = []; }
    if (source['observedSeq']) { this.observedSeq = new fhir.FhirString({value: source.observedSeq}); }
    if (source['_observedSeq']) {
      if (this.observedSeq) { this.observedSeq.addExtendedProperties(source._observedSeq!); }
      else { this.observedSeq = new fhir.FhirString(source._observedSeq as Partial<fhir.FhirStringArgs>); }
    }
    if (source['quality']) { this.quality = source.quality.map((x) => new fhir.SequenceQuality(x)); }
    else { this.quality = []; }
    if (source['readCoverage']) { this.readCoverage = new fhir.FhirInteger({value: source.readCoverage}); }
    if (source['_readCoverage']) {
      if (this.readCoverage) { this.readCoverage.addExtendedProperties(source._readCoverage!); }
      else { this.readCoverage = new fhir.FhirInteger(source._readCoverage as Partial<fhir.FhirIntegerArgs>); }
    }
    if (source['repository']) { this.repository = source.repository.map((x) => new fhir.SequenceRepository(x)); }
    else { this.repository = []; }
    if (source['pointer']) { this.pointer = source.pointer.map((x) => new fhir.Reference(x)); }
    else { this.pointer = []; }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation(exp:string = ''):fhir.FtsIssue[] {
    let issues:fhir.FtsIssue[] = super.doModelValidation(exp);
    if (exp === '') { exp = 'Sequence' }
    this.vRS('resourceType',exp)
    this.vOA('identifier',exp)
    this.vOS('type',exp)
    this.vRS('coordinateSystem',exp)
    this.vOS('patient',exp)
    this.vOS('specimen',exp)
    this.vOS('device',exp)
    this.vOS('performer',exp)
    this.vOS('quantity',exp)
    this.vOS('referenceSeq',exp)
    this.vOA('variant',exp)
    this.vOS('observedSeq',exp)
    this.vOA('quality',exp)
    this.vOS('readCoverage',exp)
    this.vOA('repository',exp)
    this.vOA('pointer',exp)
    return issues;
  }
}
